job-runner/chain/chain.go

```diff
@@ -340,16 +485,70 @@ func (c *chain) UnmarshalJSON(bytes []byte) error {
 
 	c.jcMux = &sync.RWMutex{}
 	c.jobChain = m.JobChain
-
+	c.seqTriesMux = &sync.RWMutex{}
+	c.sequenceTries = m.SequenceTries
+	c.totalTriesMux = &sync.RWMutex{}
+	c.totalTries = m.TotalJobTries
+	c.latestTriesMux = &sync.RWMutex{}
+	c.latestRunTries = m.LastRunJobTries
+	c.n = m.NumJobsRun
 	c.runningMux = &sync.RWMutex{}
 	c.running = m.Running
-	c.n = m.N
 
 	return nil
 }
 
 // -------------------------------------------------------------------------- //
 
+// jobIsReady returns whether or not a job is ready to run. A job is considered
+// ready to run if it is Pending or Stopped with some retry attempts remaining,
+// and all of its previous jobs are complete. If any previous jobs are not
+// complete, the job is not ready to run.
+// Doesn't lock jcMux, so suitable to call within other chain methods lock jcMux.
+func (c *chain) jobIsReady(jobId string) bool {
+	// Only Pending or Stopped jobs with available retries can be ready to run.
+	job := c.jobChain.Jobs[jobId]
+	switch job.State {
+	case proto.STATE_PENDING:
+	case proto.STATE_STOPPED:
```

Comment:
`case proto.STATE_PENDING, proto.STATE_STOPPED`. I suspect this doesn't work because Go switch don't fall through, so this should fail for `proto.STATE_PENDING`.